feat(profile): resolve all controls in profile

meganwolf0 opened the pull request:
## Description

Library functions to resolve all controls from a provided profile. Returns a map of source UUIDs to the respective controls. Intent of this structure is to properly map referenced source (from comp-defn) to the respective controls that are dictated by the profile.

Important Additions/Modifications:
* Refactored the extraction of controls from a provided profile in `component.go` - the addition of this PR should allow for generation of component directly from a profile, HOWEVER, note the below constraint on profiles using internal refs
* Updated network functionality to read from local files - should handle multiple path types, I think it was limited before on "file://" definitions which aren't always picked up for local paths
* Unit tests for resolving catalog and profile controls

Important Note - 
Including profiles with references to internal UUIDs (e.g., the way NIST does [this profile](https://github.com/usnistgov/oscal-content/blob/main/nist.gov/SP800-53/rev5/yaml/NIST_SP-800-53_rev5_MODERATE-baseline_profile.yaml)) will not resolve.

## Related Issue

Fixes #813 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [X] Test, docs, adr added or updated as needed
- [ ] (n/a) [Schema Updates](https://github.com/defenseunicorns/lula/blob/main/docs/community-and-contribution/schema-updates.md) applied
- [X] [Contributor Guide Steps](https://github.com/defenseunicorns/lula/blob/main/CONTRIBUTING.md) followed

brandtkeller commented on src/pkg/common/network/network.go:
```diff
@@ -163,7 +163,12 @@ func GetLocalFileDir(inputURL, baseDir string) string {
+	// - ./path/to/file
+	// - /path/to/file
+	// - https://example.com/path/to/file
+	if url.Scheme == "file" || !url.IsAbs() {
```
The docs are ambiguous - but I am curious if the `scheme` is a required part of the URI use in OSCAL.

meganwolf0 replied:
I don't think `file` prefix is required - if you look at some of the [oscal examples](https://github.com/usnistgov/oscal-content/blob/main/examples/component-definition/yaml/example-component.yaml) for fields that specify uri (see href in backmatter) those just do relative paths

brandtkeller replied:
Roger! would be good to validate this with the constraints we talked about in metaschema and oscal-cli but otherwise sounds good.

meganwolf0 commented:
Temporarily pulling this down to address a bug I found in component generate - will also update those tests accordingly.

meganwolf0 commented:
> Temporarily pulling this down to address a bug I found in component generate - will also update those tests accordingly.

Realized that the `AddControl` logic was breaking the control search loop before child controls could be checked, so needed to re-order some of that functionality. Also updated the debugging logs in `ComponentFromCatalog` to accurately reflect controls not-found. Updated the `TestComponentFromCatalog` to check for an additional control that is nested under an excluded control.

brandtkeller reviewed (approved):
Great point of iteration - I don't see anything that has to block progress but this does generate a lot of fun opportunities.
